### PR TITLE
added ability to create hard links

### PR DESCRIPTION
**Pull request by Quark-X10**
**Description**

- added a method in NodeTraits to create hard links

**How to test this?**

```bash
cmake ..
make -j8
make test
```

**Test System**
 - OS: Ubuntu 22.04.2 LTS (WSL)
 - Compiler: gcc 11.3.0
 - Dependency versions: hdf5 1.10.7

### CODE REVIEW

**Comment by alkino** [comment]
Should fix #357 ?

**Comment by Quark-X10** [comment]
I think it should fix #357
I let a typo slip into my PR, sorry for that. Actually this is my first PR ever.

**Comment by 1uc** [comment]
>  Actually this is my first PR ever.

In that case: Well done! It includes a sensible test which is very good. The only things I'd like to change is the description, since I never know which way links go. We'll fix the formatting issue. Clang-format is a bit particular about using exactly the correct version or otherwise the output differs.

**Comment by Quark-X10** [comment]
Thanks !
> We'll fix the formatting issue.

I committed the formatting change before seeing your comment sorry.

> The only things I'd like to change is the description, since I never know which way links go. 

I agree with you, I'm always confused about links direction, if "target object" and "link name" are not clear enough, feel free to use a better wording !

File: include/highfive/bits/H5Node_traits_misc.hpp [hunk 1]
```diff
@@ -358,6 +358,29 @@ inline void NodeTraits<Derivate>::createExternalLink(const std::string& link_nam
     }
 }
 
+template <typename Derivate>
+template <typename T, typename>
+inline void NodeTraits<Derivate>::createHardLink(const std::string& link_name,
+                                                 const T& target_obj,
+                                                 LinkCreateProps linkCreateProps,
```
Comment (alkino) [inline include/highfive/bits/H5Node_traits_misc.hpp:365]:
```suggestion
                                                 const LinkCreateProps& linkCreateProps,
```

Comment (Quark-X10) [reply include/highfive/bits/H5Node_traits_misc.hpp:365]: As in createSoftLink and createExternalLink, if parents is true, the link create properties have to be modified.
I tried to maintain consistent API with these 2 methods.

Comment (alkino) [reply include/highfive/bits/H5Node_traits_misc.hpp:365]: thank you for explanation :)

**Comment by 1uc** [comment]
Thank you, this looks good. I'll merge it into a branch `hard-links`, add a few additional tests. Mostly to check if I understand it correctly.

**Pull request merged**